### PR TITLE
return proper errors w/ statuscodes for repoupdater client

### PR DESCRIPTION
**Pull request by notjrbauer**
refs #9274

[This call](https://github.com/sourcegraph/sourcegraph/blob/master/cmd/frontend/backend/repos.go#L70) eventually makes its way into [this call](https://github.com/sourcegraph/sourcegraph/blob/master/cmd/frontend/backend/repos.go#L115). The error returned from `RepoLookup` does not implement `errorcode.NotFounder`, so that's my guess as to why the server was returning a 500. I checked the rest of the  repoupdater client, and discovered it was missing auth and temporary error types. I decided to try and add those too.

I can try and add a test or two if this looks alright.

### CODE REVIEW

File: cmd/frontend/backend/repos_vcs.go [hunk 1]
```diff
@@ -20,6 +20,18 @@ import (
...
+	// ErrNotFound is when a repository is not found.
+	ErrNotFound = errors.New("repository not found")
+
+	// ErrUnauthorized is when an authorization error occurred.
+	ErrUnauthorized = errors.New("not authorized")
+
+	// ErrTemporarilyUnavailable is when the repository was reported as being temporarily
+	// unavailable.
+	ErrTemporarilyUnavailable = errors.New("repository temporarily unavailable")
+)
```
Comment (unknwon) [inline cmd/frontend/backend/repos_vcs.go:33]: Can you elaborate what problem is solved by moving this block from `internal/repoupdater/client.go` to here? Why not/can't reuse the new error types defined in this PR?

Comment (notjrbauer) [reply cmd/frontend/backend/repos_vcs.go]: @unknwon you were right to be concerned - I made an update that reuses the new error types and how we determine their types

**Comment by unknwon** [comment]
CI failed with:
```
FAIL: TestRepoLookup (0.10s) | 0s
-- | --
  | --- FAIL: TestRepoLookup/not_found (0.00s)
  | server_test.go:1106: have err: "repo Not found (name=github.com/a/b  notfound=true) ", want: "repository not found"
  | --- FAIL: TestRepoLookup/not_found_-_GitHub.com_on_Sourcegraph.com (0.00s)
  | server_test.go:1106: have err: "repo Not found (name=github.com/foo/bar  notfound=true) ", want: "repository not found"
  | --- FAIL: TestRepoLookup/unauthorized_-_GitHub.com_on_Sourcegraph.com (0.01s)
  | server_test.go:1106: have err: "repo access Not authorized (name=github.com/foo/bar  noauthz=true) ", want: "not authorized"
  | --- FAIL: TestRepoLookup/temporarily_unavailable_-_GitHub.com_on_Sourcegraph.com (0.00s)
  | server_test.go:1106: have err: "repo temporary unavailable (name=github.com/foo/bar  istemporary=true) ", want: "repository temporarily unavailable"
  | --- FAIL: TestRepoLookup/GithubDotcomSource_on_Sourcegraph.com_ignores_non-Github.com_repos (0.00s)
  | server_test.go:1106: have err: "repo Not found (name=git-codecommit.us-west-1.amazonaws.com/stripe-go  notfound=true) ", want: "repository not found"
```

**Comment by notjrbauer** [comment]
updated tests

alt

File: cmd/frontend/internal/app/ui/handlers.go [hunk 2]
```diff
@@ -166,12 +166,12 @@ func newCommon(w http.ResponseWriter, r *http.Request, title string, serveError
...
 				return nil, nil
 			}
-			if errcode.IsNotFound(err) || errors.Cause(err) == repoupdater.ErrNotFound {
+			if errcode.IsNotFound(err) {
 				// Repo does not exist.
 				serveError(w, r, err, http.StatusNotFound)
 				return nil, nil
 			}
-			if errors.Cause(err) == repoupdater.ErrUnauthorized {
+			if errcode.IsUnauthorized(err) {
```
Comment (unknwon) [inline cmd/frontend/internal/app/ui/handlers.go:174]: Hmm, have you verified this PR will make a difference on the behavior? It looks like we're already checking `repoupdater.ErrNotFound`?

Comment (notjrbauer) [reply cmd/frontend/internal/app/ui/handlers.go:174]: I'll write some tests and open this back up once I can demonstrate.

Comment (notjrbauer) [reply cmd/frontend/internal/app/ui/handlers.go:174]: I couldn't leverage any existing tests but I found a concrete case of 500 vs 404 in the issues.

Ref #10000
uri: https://sourcegraph.test:3443/code.gitea.io/sdk/gitea/-/badge.svg

[This http middleware](https://github.com/sourcegraph/sourcegraph/blob/master/cmd/frontend/internal/app/app.go#L38) assumes the returned err adheres to the errcode interface when determining statuscodes.  

[You can see the implementation of the case here](https://github.com/sourcegraph/sourcegraph/blob/master/internal/errcode/code.go#L53-L56)

My proposed changes regarding repo updater errors allows the server to return a proper 404

Comment (notjrbauer) [reply cmd/frontend/internal/app/ui/handlers.go:174]: Update: I believe anything using the `errorutil.Handler` will return 5xxs if it returns a repository not found error && the error does not implement the NotFounder interface.
```
৸ ag errorutil.Handler
internal/app/app.go
38:     r.Get(router.RepoBadge).Handler(trace.TraceRoute(errorutil.Handler(serveRepoBadge)))
50:             r.Get(router.GoSymbolURL).Handler(trace.TraceRoute(errorutil.Handler(serveGoSymbolURL)))
65:     r.Get(router.GDDORefs).Handler(trace.TraceRoute(errorutil.Handler(serveGDDORefs)))
66:     r.Get(router.Editor).Handler(trace.TraceRoute(errorutil.Handler(serveEditor)))

internal/app/debugproxies/handler.go
38:     r.PathPrefix("/proxies").Handler(http.StripPrefix("/-/debug/proxies", adminOnly(errorutil.Handler(rph.serveReverseProxy))))
```

[Patch of test showing 500s](https://github.com/notjrbauer/sourcegraph/commit/aecec8f012ae84956c4a14d62db9964a016f8fae.patch)

Comment (unknwon) [reply cmd/frontend/internal/app/ui/handlers.go:174]: Thanks! This makes sense!

**Comment by unknwon** [comment]
CI failed:
```
go test | 1m 34s
-- | --
  | ?   	github.com/sourcegraph/sourcegraph	[no test files]
  | # github.com/sourcegraph/sourcegraph/cmd/frontend/internal/app/ui [github.com/sourcegraph/sourcegraph/cmd/frontend/internal/app/ui.test]
  | cmd/frontend/internal/app/ui/handlers_test.go:101:20: type repoupdater.ErrNotFound is not an expression
  | cmd/frontend/internal/app/ui/handlers_test.go:106:20: type repoupdater.ErrUnauthorized is not an expression
```

**Comment by notjrbauer** [comment]
ahhh apologies -  I rebased on my older branch

**Pull request closed without merging**